skmonitor/src/Monitor.sk

```diff
@@ -0,0 +1,398 @@
+      Event::create(
+        SKStore.duplicate(name),
+        unsafe_chill_trust_me(attributesCopy),
+      ),
+    );
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setOkStatus(): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SOk());
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setErrorStatus(description: String = ""): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SError(), SKStore.duplicate(description));
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setUnsetStatus(): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SUnset());
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun endSpan(): void {
+  monitoring() match {
+  | MOn _ ->
+    get_span().each(span -> {
+      SKStore.switchToTrace();
+      span.!optEnd = Some(now());
+      if (span.optParent.isNone()) {
+        memory = memory();
+        span.setIntAttribute("skruntime.memory.persistent", memory.persistent);
+        span.setIntAttribute("skruntime.memory.freetable", memory.freetable);
+        span.setIntAttribute("skruntime.obstack.pick", memory.obstack_pick);
+      };
+      impact = get_trace_size(span.optParent.map(p -> p.i1));
+      span.setIntAttribute("skmonitor.memory", impact);
+      flush(span);
+      nSpan = span.optParent match {
+      | Some((parent, saved)) ->
+        SKStore.destroyObstack(saved);
+        Some(parent)
+      | _ -> None()
+      };
+      set_span(nSpan);
+      if (nSpan.isNone()) {
+        set_writer(None());
+        SKStore.clearAndRestoreFromTrace()
+      } else {
+        SKStore.restoreFromTrace()
+      }
+    })
+  | MMemory _ -> print_debug(memory().toJSON().toString())
+  | MNo _ -> void
+  }
+}
+
+fun format(msg: String, meta: Logger.Metadata): String {
+  `${Path.basename(meta.filename)}:${meta.lineno} - ${msg}`
+}
+
+fun log(level: Level, msg: String, meta: Logger.Metadata): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.addEvent(
+      Event::create(
+        "sklogger.log",
+        Map[
+          "sklogger.level" => VString(level.toString()),
+          "sklogger.message" => VString(format(msg, meta)),
+        ],
+      ),
+    );
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun trace(msg: String, meta: Logger.Metadata): void {
+  log(Trace(), msg, meta)
+}
+
+fun debug(msg: String, meta: Logger.Metadata): void {
+  log(Debug(), msg, meta)
+}
+
+fun info(msg: String, meta: Logger.Metadata): void {
+  log(Info(), msg, meta)
+}
+
+fun warn(msg: String, meta: Logger.Metadata): void {
+  log(Warn(), msg, meta)
+}
+
+fun error(msg: String, meta: Logger.Metadata): void {
+  log(Error(), msg, meta)
+}
+
+fun fatal(msg: String, meta: Logger.Metadata): void {
+  log(Fatal(), msg, meta)
+}
+
+fun flush(span: readonly Span): void {
+  trace = span.toJSON().toString();
+  get_writer() match {
+  | Some(writer) -> writer(`${trace}\n`)
```

Comment:
This will copy the string right? Can we call writer twice? There might be other opportunities like this.